### PR TITLE
Reduction class for single-crystal diffraction data from SXD and WISH instruments

### LINKED ISSUE #34217: Update WISH single crystal reduction class
+ discuss\demo current state to WISH team

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

Adds a base single-crystal reduction class (`BaseSX`) and an inherited class for instrument specific operation for WISH (`WishSX` and SXD (`SXD`). 

In terms of tests:

- Static functions in the base class (`BaseSX` and `SXD`) have been unit tested
- SXD systems test have been re-written to use the new class and significantly added to (this tests all the non static functions and the whole workflow)
- WISH systems test for the vanadium processing now uses the `WishSX` class

These classes are a work in progress, in particular the WISH one  - but it's important to merge this into the code base so that it can be tested by scientists during the cycle, and to unite lots of helper functions that lived in different scripts. Also to stop the two workflows (SXD and WISH evolving separately such that e.g. bug fixes being made to one script but not the other etc.)

**To test:**

(1) Run a detector calibration

```
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from Diffraction.single_crystal.base_sx import PEAK_TYPE, INTEGRATION_TYPE
from Diffraction.single_crystal.sxd import SXD

sxd = SXD()
wsname = sxd.load_run(23767)
ClearInstrumentParameters(Workspace=wsname)
pk_wsname = sxd.find_sx_peaks(wsname,  nstd=8)
sxd.remove_peaks_on_detector_edge(pk_wsname, 2)

# inital UB
FindUBUsingLatticeParameters(PeaksWorkspace=pk_wsname , a=5.6402, b=5.6402, c=5.6402, 
                             alpha=90, beta=90, gamma=90, FixParameters=False)
IndexPeaks(PeaksWorkspace=pk_wsname, Tolerance=0.15, CommonUBForAll=True)
# force cubic (does better job than FindUBUsingLatticeParameters with FixParameters=True)
CalculateUMatrix(PeaksWorkspace=pk_wsname , a=5.6402, b=5.6402, c=5.6402, 
                 alpha=90, beta=90, gamma=90)
IndexPeaks(PeaksWorkspace=pk_wsname, Tolerance=0.15, CommonUBForAll=False) # 229/290

save_dir = r"C:\Users\xhg73778\Desktop"
detcal_path = sxd.calibrate_sxd_panels(wsname, pk_wsname, save_dir, tol=0.25)

IndexPeaks(PeaksWorkspace=pk_wsname, Tolerance=0.15, CommonUBForAll=True)  # 262/290
```

(2) Run an NaCl reduction (using the `detcal_path` you made in step 1)

```
save_dir = r"C:\Users\xhg73778\Desktop"
detcal_path = r"C:\Users\xhg73778\Desktop\SXD23767_peaks_detcal.xml"
# integration args
intPeaksMDArgs = {'ellipsoid': False}
# setup class                    
sxd = SXD(vanadium_runno=32671, empty_runno=32673, detcal_path=detcal_path)
sxd.set_sample(Geometry={'Shape': 'CSG', 'Value': sxd.sphere_shape}, 
               Material={'ChemicalFormula': 'Na Cl', 'SampleNumberDensity': 0.0223})
sxd.set_goniometer_axes([0,1,0,1])  # ccw rotation around vertical
sxd.process_vanadium()

# process a single run
for run in range(33335,33336):
    sxd.process_data([run], [0]) # 0 gonio rotation
    sxd.convert_to_MD(run=run)  # convert to QLab (default)
    # find peaks
    ws = sxd.get_ws(run)
    peaks = sxd.find_sx_peaks(ws,  nstd=12)
    sxd.remove_peaks_on_detector_edge(peaks, 2)
    sxd.remove_duplicate_peaks_by_qlab(peaks, q_tol=0.05)
    sxd.set_peaks(run, peaks)
    # Find a UB (or load one etc. - will rotate UB by goniometer)
    FindUBUsingFFT(PeaksWorkspace=peaks, MinD=1, MaxD=10)
    # calibrate sample position
    sxd.calibrate_sample_pos(tol=0.15)
    # predict peaks
    sxd.predict_peaks(MinDSpacing=0.5, WavelengthMin=0.4, ReflectionCondition='All-face centred')
    # integrate peaks
    sxd.integrate_data(INTEGRATION_TYPE.MD_OPTIMAL_RADIUS, PEAK_TYPE.FOUND, 
                                  run=run, scale=12, **intPeaksMDArgs)
   sxd.save_peak_table(run, PEAK_TYPE.FOUND, INTEGRATION_TYPE.MD_OPTIMAL_RADIUS, save_dir, save_format='Jana')
```
   
(3) Check systems tests pass

Fixes #34217

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Diffraction/single_crystal/sxd.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+        mantid.RotateInstrumentComponent(Workspace=wsname, ComponentName="bank1", X=1, Y=1, Angle=180, EnableLogging=False)
+        mantid.RotateInstrumentComponent(Workspace=wsname, ComponentName="bank1", X=1, Angle=-37.5, EnableLogging=False)
+        mantid.MoveInstrumentComponent(
+            Workspace=wsname, ComponentName="bank1", X=0.136971, Z=-0.178505, RelativePosition=False, EnableLogging=False
+        )
+
+    @staticmethod
+    def calibrate_sxd_panels(ws, peaks, save_dir, tol=0.15, maxShiftInMeters=0.025):
+        """
+        Calibrate SXD panels and apply the calibration to th
```
Comment (DannyHindson) [inline scripts/Diffraction/single_crystal/sxd.py:130]: Given this is a fairly chunky function it's worth filling out the description here

File: scripts/Diffraction/single_crystal/base_sx.py [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+                func(self, *args, **kwargs)
+            else:
+                for run in self.runs.keys():
+                    func(self, *args, **kwargs, run=run)
+
+        return apply
+
+    # --- getters ---
+
+    def get_ws(self, run):
```
Comment (DannyHindson) [inline scripts/Diffraction/single_crystal/base_sx.py:47]: Does this return a ws name? Might be better to call it `get_ws_name` if yes. Similar comment for `get_md`

File: scripts/Diffraction/single_crystal/sxd.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+                FilterVariable="RunNumber",
+                BankName="bank" + str(ibank),
+                EnableLogging=False,
+            )
+            # report inital indexing
+            npeaks_bank = SXD.retrieve(peaks_bank).getNumberPeaks()
+            nindexed, *_ = mantid.IndexPeaks(PeaksWorkspace=peaks_bank, Tolerance=tol_to_report, CommonUBForAll=False, EnableLogging=False)
+            logger.notice(f"Indexed before optimization = {nindexed}/{npeaks_bank} peaks within {tol_to_report}")
+            mantid.IndexPeaks(PeaksWorkspace=peaks_name, Tolerance=tol, CommonUBForAll=False, EnableLogging=False)
+            peaks_bank_mod, pos_table, *_ = mantid.OptimizeCrystalPlacement(
```
Comment (DannyHindson) [inline scripts/Diffraction/single_crystal/sxd.py:194]: After running the test SXD script I see a couple of workspaces left behind called CovarianceInfo or FitInfoTable which seem to be from OptimizeCrystalPlacement. If these are from the last bank processed they perhaps don't have any special significance. So could these be deleted?

File: scripts/Diffraction/single_crystal/base_sx.py [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+            axis_dict = {f"Axis{iax}": ",".join([str(angles[iax]), self.gonio_axes[iax]]) for iax in range(len(angles))}
+            mantid.SetGoniometer(Workspace=ws, EnableLogging=False, **axis_dict)
+
+    @staticmethod
+    def convert_ws_to_MD(wsname, md_name=None, frame="Q (lab frame)"):
+        if md_name is None:
+            md_name = wsname + "_MD"
+        xunit = BaseSX.get_xunit(wsname)  # get initial xunit
+        BaseSX._normalise_by_bin_width_in_k(wsname)  # normalise by bin-width in K = 2pi/lambda
+        wsMD = mantid.ConvertToDiffractionMDWorkspace(
```
Comment (DannyHindson) [inline scripts/Diffraction/single_crystal/base_sx.py:148]: Does this call leave a PreprocessedDetectorsWS behind? Might be worth cleaning it up at the end of the function

File: scripts/Diffraction/wish/wishSX.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+import mantid.simpleapi as mantid
+from Diffraction.single_crystal.base_sx import BaseSX
+from mantid.api import AnalysisDataService as ADS
+
+
+class WishSX(BaseSX):
+    def __init__(self, vanadium_runno=None):
+        super().__init__(vanadium_runno)
+        self.grp_ws = None  # banks grouping workspace
+        self.ngrp = None  # no. groups (i.e. number of banks)
```
Comment (DannyHindson) [inline scripts/Diffraction/wish/wishSX.py:12]: are `self.grp_ws` and `self.ngrp` ever used in this class?

File: scripts/Diffraction/single_crystal/sxd.py [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+        return wsname
+
+    def _replace_spectra_with_focussed_bank(self, wsname):
+        # important to focus in TOF not d-spacing (lambda would be better but not compatible with Diffractionfocussing)!
+        ws = SXD.retrieve(wsname)
+        banks_foc = mantid.GroupDetectors(
+            InputWorkspace=ws,
+            OutputWorkspace="grouped",
+            IgnoreGroupNumber=False,
+            CopyGroupingFromWorkspace=self.grp_ws,
```
Comment (DannyHindson) [inline scripts/Diffraction/single_crystal/sxd.py:71]: Perhaps the grp_ws could just be passed into this function as a parameter. Doesn't seem like it's used anywhere else so having as a field on the class may be unnecessary

File: scripts/Diffraction/wish/wishSX.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+                    self._set_goniometer_on_ws(wsname, gonio_angles)
+                else:
+                    # gonio_angles is a list of individual or tuple motor angles for each run
+                    self._set_goniometer_on_ws(wsname, gonio_angles[irun])
+            # correct for empty counts and normalise by vanadium
+            self._divide_workspaces(wsname, self.van_ws)
+            # set sample (must be done after gonio to rotate shape) and correct for attenuation
+            if self.sample_dict is not None:
+                mantid.SetSample(wsname, EnableLogging=False, **self.sample_dict)
+                mantid.ConvertUnits(InputWorkspace=wsname, OutputWorkspace=wsname, Target="Wavelength", EnableLogging=False)
```
Comment (DannyHindson) [inline scripts/Diffraction/wish/wishSX.py:37]: Similar comment to the one in the sxd.py unit - are the units on the ws at the end of this function different depending on whether the absorption correction is done. Think the final units are wavelength if code comes in here

File: scripts/Diffraction/wish/wishSX.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        self._divide_workspaces(self.van_ws, vanadium_transmission)
+        mantid.DeleteWorkspace(vanadium_transmission)
+
+    @staticmethod
+    def load_run(runno):
+        wsname = "WISH000" + str(runno)
+        ws, mon = mantid.LoadRaw(Filename=wsname + ".raw", OutputWorkspace=wsname, LoadMonitors="Separate")
+        mon_name = mon.name()
+        mantid.CropWorkspace(InputWorkspace=wsname, OutputWorkspace=wsname, XMin=6000, XMax=99000)
+        mantid.CropWorkspace(InputWorkspace=mon_name, OutputWorkspace=mon_name, XMin=6000, XMax=99000)
```
Comment (DannyHindson) [inline scripts/Diffraction/wish/wishSX.py:88]: Similar comment to sxd.py. I think you should pull out these crop ranges as constants

**Review by DannyHindson — DISMISSED** [review]
I've had an initial look around. It's obviously been a lot of work pulling this together so well done - looks a very nice job.
I've added a few pretty trivial comments but these points aside all seems to work well
The second test script on this PR should have a call to `remove_peaks_on_detector_edge` instead of `filter_peaks_on_detector_edge` so worth updating that in case people use this PR as a reference

File: scripts/Diffraction/single_crystal/base_sx.py (hunk 2 above)
Comment (RichardWaiteSTFC) [reply scripts/Diffraction/single_crystal/base_sx.py:53]: > Does this return a ws name? Might be better to call it `get_ws_name` if yes. Similar comment for `get_md`

I would like to encourage users to use strings in their scripts (less chance of bugs through an object becoming invalidated, and easier/more transparent when assigning the output of an algorithm to have the same name) so I force the class setters to set the internal dict values with workspace names. That said the user can access the internal dict (a downside of python...) and set the workspace to a python object if they wanted - and it's designed so everything will (should!) work anyway! i.e. I assume the return from a getter or function input could be a string or python object  - hence the `BaseSX.retrieve` method is called a lot. This makes it a bit more cumbersome for the developer but I think should be more robust...

So I would say it is a getter of the dict field `ws` - but I wouldn't want to bake in an assumption as to the data type - but maybe that's just confusing? Or I should have both?

Comment (RichardWaiteSTFC) [reply scripts/Diffraction/single_crystal/base_sx.py:53]: Yep I've thought about this a bit more and definitely need to be clear what data type the getters return - I will force it to be a string and call put `name` in the function name. Not sure I will add getters for ws objects as well, will rely on BaseSX.retrieve if users want that for now (hopefully they won't need it now there's the helper functions)

**Comment by sf1919** [comment]
@RichardWaiteSTFC  As there are what appears to be valid test failures we won't re-run the ones that are currently 'scheduled' as they will rerun when you push a new commit

**Comment by thomashampson** [comment]
pre-commit.ci run

**Review by jhaigh0 — APPROVED** [review]
Can approve now the tests have been fixed

**Pull request merged**